packages/teleterm/src/ui/uri.ts

```diff
@@ -30,6 +30,54 @@ export const paths = {
   doc: '/docs/:docId',
 };
 
+type RootClusterId = string;
+type LeafClusterId = string;
+type ServerId = string;
+type KubeId = string;
+type DbId = string;
+export type RootClusterUri = `/clusters/${RootClusterId}`;
+export type RootClusterServerUri =
+  `/clusters/${RootClusterId}/servers/${ServerId}`;
+export type RootClusterKubeUri = `/clusters/${RootClusterId}/kubes/${KubeId}`;
+export type RootClusterDatabaseUri = `/clusters/${RootClusterId}/dbs/${DbId}`;
+export type RootClusterResourceOnlyUri =
+  | RootClusterServerUri
+  | RootClusterKubeUri
+  | RootClusterDatabaseUri;
+export type RootClusterResourceUri =
+  | RootClusterUri
+  | RootClusterResourceOnlyUri;
+export type LeafClusterUri =
+  `/clusters/${RootClusterId}/leaves/${LeafClusterId}`;
+export type LeafClusterServerUri =
+  `/clusters/${RootClusterId}/leaves/${LeafClusterId}/servers/${ServerId}`;
+export type LeafClusterKubeUri =
+  `/clusters/${RootClusterId}/leaves/${LeafClusterId}/kubes/${KubeId}`;
+export type LeafClusterDatabaseUri =
+  `/clusters/${RootClusterId}/leaves/${LeafClusterId}/dbs/${DbId}`;
+export type LeafClusterResourceOnlyUri =
+  | LeafClusterServerUri
+  | LeafClusterKubeUri
+  | LeafClusterDatabaseUri;
+export type LeafClusterResourceUri =
+  | LeafClusterUri
+  | LeafClusterResourceOnlyUri;
```

Comment:
There's lots of repetition here but this stuff is unlikely to change over time any way. It's not like we're going to add a third type of a cluster.

I'm not sure about the `*ResourceOnlyUri` name, but I basically want to be able to say "I accept URIs to resources only" and "I accept URIs to resources and clusters as well".

Maybe `*ResourceOnlyUri` should be renamed to `*ResourceUri` and current `*ResourceUri` to `*ClusterOrResourceUri`.

Reply:
> Maybe *ResourceOnlyUri should be renamed to *ResourceUri and current *ResourceUri to *ClusterOrResourceUri.

I agree. Currently, when we refer to a _resource_ we think about servers/dbs/kubes, not clusters.

Reply:
repetition is fine with me if it helps us down the road. this is REALLY cool btw. I didn't know that template literals could be used this way. This PR taught me a lot. 

+1 for _ClusterOrResourceUri_